raft/src/test/java/org/apache/kafka/raft/ConnectionCacheTest.java

```diff
@@ -0,0 +1,261 @@
+            new InetSocketAddress("127.0.0.1", 9094),
+            1L
+        ));
+
+        // One host has an inflight request
+        ConnectionCache.ConnectionState connectionState1 = cache.getOrCreate(1);
+        connectionState1.onRequestSent(1, time.milliseconds());
+        assertFalse(connectionState1.isReady(time.milliseconds()));
+
+        // Another is backing off
+        ConnectionCache.ConnectionState connectionState2 = cache.getOrCreate(2);
+        connectionState2.onRequestSent(2, time.milliseconds());
+        connectionState2.onResponseError(2, time.milliseconds());
+        assertFalse(connectionState2.isReady(time.milliseconds()));
+
+        cache.resetAll();
+
+        // Now both should be ready
+        assertTrue(connectionState1.isReady(time.milliseconds()));
+        assertTrue(connectionState2.isReady(time.milliseconds()));
+    }
+
+    @Test
+    public void testBackoffAfterFailure() {
+        List<InetSocketAddress> bootstrapServers = Collections.singletonList(
+            new InetSocketAddress("127.0.0.1", 9092)
+        );
+
+        ConnectionCache cache = new ConnectionCache(networkChannel,
+            bootstrapServers,
+            retryBackoffMs,
+            requestTimeoutMs,
+            logContext);
+
+        ConnectionCache.HostInfo hostInfo = new ConnectionCache.HostInfo(
+            new InetSocketAddress("127.0.0.1", 9093),
+            1L
+        );
+
+        ConnectionCache.ConnectionState connectionState = cache.getOrCreate(1);
+        assertFalse(connectionState.isReady(time.milliseconds()));
+
+        cache.maybeUpdate(1, hostInfo);
+        assertTrue(connectionState.isReady(time.milliseconds()));
+
+        long correlationId = 1;
+        connectionState.onRequestSent(correlationId, time.milliseconds());
+        assertFalse(connectionState.isReady(time.milliseconds()));
+
+        connectionState.onResponseError(correlationId, time.milliseconds());
+        assertFalse(connectionState.isReady(time.milliseconds()));
+
+        time.sleep(retryBackoffMs);
+        assertTrue(connectionState.isReady(time.milliseconds()));
+    }
+
+    @Test
+    public void testSuccessfulResponse() {
+        List<InetSocketAddress> bootstrapServers = Collections.singletonList(
+            new InetSocketAddress("127.0.0.1", 9092)
+        );
+
+        ConnectionCache cache = new ConnectionCache(networkChannel,
+            bootstrapServers,
+            retryBackoffMs,
+            requestTimeoutMs,
+            logContext);
+
+        ConnectionCache.HostInfo hostInfo = new ConnectionCache.HostInfo(
+            new InetSocketAddress("127.0.0.1", 9093),
+            1L
+        );
+
+        ConnectionCache.ConnectionState connectionState = cache.getOrCreate(1);
+        cache.maybeUpdate(1, hostInfo);
+
+        long correlationId = 1;
+        connectionState.onRequestSent(correlationId, time.milliseconds());
+        connectionState.onResponseReceived(correlationId);
```

Comment:
nit: could we `assertFalse(connectionState.isReady(time.milliseconds()));` before `onResponseReceived` as well.